### PR TITLE
Lowdiff shares (2nd update #2718)

### PR DESCRIPTION
**Pull request by r4sas**
Please verify and test before merging.

### CODE REVIEW

**Review by TheSerapher — COMMENTED** [review]
Check my change comment, other than that it looks okay. Make sure to test it a bit

File: include/classes/statistics.class.php [hunk 1]
```diff
@@ -575,9 +575,9 @@ public function getUserMiningStats($username, $account_id=NULL, $interval=180) {
     if ($this->getGetCache() && $data = $this->memcache->get(__FUNCTION__ . $account_id)) return $data;
     $stmt = $this->mysqli->prepare("
       SELECT
-        IFNULL(SUM(difficulty) / ?, 0) AS sharerate,
-        IFNULL(SUM(difficulty), 0) AS shares,
-        IFNULL(AVG(difficulty), 0) AS avgsharediff
+        IFNULL(SUM(difficulty) / ?, " . $this->coin->getShareDifficultyPrecision() . ") AS sharerate,
```
[the remark this answers is not in the document]
Comment (r4sas) [reply include/classes/statistics.class.php]: Ok, will change back

**Comment by TheSerapher** [comment]
Once merged maybe @HashUnlimited can update their branch and re-run the DB schema migration. Don't forget to reset the DB_VERSION string again so it will run properly. Blocks should then show the proper amount of shares even with blocks < 1 diff share.

File: upgrade/definitions/1.0.2_to_1.0.3.inc.php [hunk 2]
```diff
@@ -9,6 +9,7 @@ function run_103() {
   $db_version_now = $setting->getValue('DB_VERSION');  // Our actual version installed
 
   // Upgrade specific variables
+  $aSql[] = "ALTER TABLE `blocks` CHANGE `shares` `shares` DOUBLE UNSIGNED DEFAULT NULL;";
   $aSql[] = "UPDATE `statistics_shares` SET `valid` = '0' WHERE `valid` IS NULL;";
   $aSql[] = "UPDATE `statistics_shares` SET `pplns_valid` = '0' WHERE `pplns_valid` IS NULL;";
   $aSql[] = "ALTER TABLE `statistics_shares` CHANGE `valid` `valid` FLOAT NOT NULL DEFAULT '0', CHANGE `invalid` `invalid` FLOAT NOT NULL DEFAULT '0', CHANGE `pplns_valid` `pplns_valid` FLOAT NOT NULL DEFAULT '0', CHANGE `pplns_invalid` `pplns_invalid` FLOAT NOT NULL DEFAULT '0';";
```
Comment (r4sas) [inline upgrade/definitions/1.0.2_to_1.0.3.inc.php]: Maybe here we must do same, and change to unsigned double?

**Comment by TheSerapher** [comment]
@HashUnlimited @r4sas since you are both kinda doing some P2P coding and testing here, please let me know when this is ready for merge into development. I think it's close to ready if not completed.

I have looked into the float/double for shares and I am not entirely sure ... I doubt think that double precision on share difficulties per user (which is what it is in that table) would not exceed that many shares. But if we want to play save we could increase it to double (but potentially waste 4 bytes per entry since it's not used).

No idea what's best? What do you think?

**Comment by r4sas** [comment]
In that case we can set to unsigned float because here will never be negative values.

**Comment by TheSerapher** [comment]
Good compromise :) Doubles the range and should be sufficient for single-account share counting.

**Comment by HashUnlimited** [comment]
sufficient for sure, the maximum diff in all bitcoin based code is 2.6959947e+67

if we are confident we aren't running into

> Using FLOAT might give you some unexpected problems because all calculations in MySQL are done with double precision.

I'm fine

**Comment by r4sas** [comment]
> Totals in the block statistics are showing integers

Let me check...

**Comment by TheSerapher** [comment]
Let's merge this now, looks complete and can be further tested on development. If more changes are required, we can start new PRs.

**Comment by r4sas** [comment]
I hope that's all :D

**Comment by HashUnlimited** [comment]
agreed, clicking through there's only one place left which is the graph but we don't want to be over-accurate here. for my part, all ready to test it squeeze it and try to break it ;-)

**Comment by r4sas** [comment]
In Statitics -> Graphs -> sharerate?

**Comment by HashUnlimited** [comment]
exactly, that's the very last place I could find :D

**Comment by r4sas** [comment]
Found where it happens. By the way I have opinion about that graphics:
https://github.com/MPOS/php-mpos/blob/e79b558163ba8eafe1964f5199f09cde74a5ebc2/templates/bootstrap/statistics/graphs/default.tpl#L5
We store that json string triple times. I think better to save it in one variable and reuse from it rather then store 3 times.

**Comment by TheSerapher** [comment]
Probably a good idea ... If you find a quick and easy way to do it, go ahead :)

**Comment by brainquan** [comment]
hello TheSerapher, i need paid support

**Comment by Hamidreza5242** [comment]
> Please verify and test before merging.

**Pull request merged**